src/ql/pass/opt/const_prop/detail/propagate.cc

```diff
@@ -0,0 +1,224 @@
+
+
+class ConstantPropagator : public ir::RecursiveVisitor {
+public:
+
+    /**
+     * Constructs a ConstantPropagator.
+     */
+    explicit ConstantPropagator(const ir::Ref &ir) : ir(ir) {
+        register_functions();
+    };
+
+private:  // RecursiveVisitor overrides
+
+    /**
+     * Fallback function.
+     */
+    void visit_node(ir::Node &node) override {
+        // Note that in our case it is not an error to come here, since we don't intend to handle all
+        // node types
+        DEBUG("visiting node '"  << ir::describe(node) << "'");
+    }
+
+    /**
+     * Visitor function for `SetInstruction` nodes.
+     */
+
+    void visit_set_instruction(ir::SetInstruction &set_instruction) override {
+        handle_expression(set_instruction.rhs);
+    }
+
+    /**
+     * Visitor function for `DynamicLoop` nodes.
+     */
+
+    void visit_dynamic_loop(ir::DynamicLoop &dynamic_loop) override {
+        handle_expression(dynamic_loop.condition);
+
+        // descend loop body
+        dynamic_loop.body->visit(*this);
+    }
+
+    /**
+     * Visitor function for `IfElseBranch` nodes.
+     */
+    void visit_if_else_branch(ir::IfElseBranch &if_else_branch) override {
+        handle_expression(if_else_branch.condition);
+
+        // descend branch body
+        if_else_branch.body->visit(*this);
+    }
+
+    /**
+     * Visitor function for `FunctionCall` nodes.
+     */
+    void visit_function_call(ir::FunctionCall &function_call) override {
+        DEBUG("recursing into operands of function call '" << ir::describe(function_call) << "'");
+        for (auto &operand : function_call.operands) {
+            handle_expression(operand);
+        }
+    }
+
+    /**
+     * Visitor function for `FunctionCall` nodes.
+     */
+    void visit_expression(ir::Expression &expression) override {
+        if(auto function_call = expression.as_function_call()) {
+            DEBUG("recursing into function call '" << ir::describe(*function_call) << "'");
+            visit_function_call(*function_call);   // descend
+        }
+    }
+
+private:    // types
+
+    // function pointer
+    typedef FncRet (*tOpFunc)(const ir::Ref &ir, const FncArgs &args);
+
+    // map function key (see register_functions()) to function pointer
+    using FuncMap = std::map<utils::Str, tOpFunc>;
+
+private:    // vars
+    const ir::Ref &ir;
+    FuncMap func_map;
+
+private:    // functions
+
+    /*
+     * Handle an expression node, i.e. replace eligible functions calls with a literal expression.
+     *
+     * Note that we cannot directly use visit_expression(), because that has a 'ir::Expression &' as parameter.
+     * It is therefore not possible to change a function_call into a (say) int_literal. Here the expression is
+     * wrapped in a 'utils::One', allowing polymorphic access. This does force us however to visit all relevant node
+     * types that contain an expression (although we can skip e.g. custom_instruction.operands).
+     */
```

Comment:
I'm still puzzling on this. This need requires a particular convention (all those '*'s) which is annoying. Another kind of visitor that visits the nodes that contain expressions would avoid this; it would also make deleting nodes easier.
And also the cqasm reader and cqasm writer are not written with a visitor, illustrating to me that the current visitor is not powerful enough. So I'd expect a family of visitors of which the current visitor is the simplest one, the visitor we actually require here is the one-but-simplest, and the cqasm reader one of the hardest. In CoSy we also had context visitors (where the visit stack was accessible as a parameter), which allowed to visit depending on attributes of nodes higher up in the IR-tree (e.g. top if-then-else, or if-then-else inside another one or inside a loop).
Have you considered changing the visit function instead of applying the current solution?

Reply:
There are multiple ways to traverse the tree depending on needs, see https://tree-gen.readthedocs.io/en/latest/tree-gen.gen.html#tree-traversal, so generally speaking it's not necessarily a bad thing that the cQASM reader/writer don't use a visitor.

I'm not sure whether the need we have here could easily be generalized: one could envision a visitor that doesn't dereference One/Any/etc, but in other places than Expressions that may not be what you want.